lib/pedicel/ec.rb

```diff
@@ -112,19 +114,19 @@ def self.symmetric_key(merchant_id:, shared_secret:)
       sha256.digest
     end
 
-    def self.merchant_id(certificate:)
+    def self.merchant_id(certificate:, config: Pedicel.config)
       begin
         cert = OpenSSL::X509::Certificate.new(certificate)
       rescue => e
         raise CertificateError, "invalid PEM format of certificate: #{e.message}"
       end
 
       merchant_id_hex =
-        cert.
-          extensions.
-          find { |x| x.oid == Pedicel.config[:oids][:merchant_identifier_field] }&.
-          value&. # Hex encoded Merchant ID plus perhaps extra non-hex chars.
-          delete("^[0-9a-fA-F]") # Remove non-hex chars.
+        cert
+        .extensions
```

Comment:
There are good reasons for both:
* https://github.com/bbatsov/ruby-style-guide#consistent-multi-line-chains
* https://github.com/bbatsov/ruby-style-guide/pull/176

I'm 🤷‍♂️ for which to choose (recently perhaps leaning towards the reasoning in https://github.com/bbatsov/ruby-style-guide/pull/176#issuecomment-20421523). I expect (because `git blame` 😛) that @kse-clearhaus prefer the "dot in the front" style (which, depending on your way of thinking, could make more sense when used with `&` as the safe navigation operator). @kse-clearhaus can you confirm?

Reply:
I fell into a crater.

Think about

```ruby
ary = [1,2,3]
```

or alike through this.

If something is commented out, then it's because it doesn't work.

# Comments

```ruby
#ary \ # Single-line comment.
#  .slice(5..-1) \
#  &.first \
#  &.zero?

ary \
  # Single-line comment.
  .slice(5..-1) \
  &.first \
  &.zero?

#ary \
#  # Multi-line
#  # comment.
#  .slice(5..-1) \
#  &.first \
#  &.zero?

ary # Single-line comment.
  .slice(5..-1)
  &.first
  &.zero?

ary
  # Single-line comment.
  .slice(5..-1)
  &.first
  &.zero?

ary.
  # Single-line comment.
  slice(5..-1)&.
  first&.
  zero?

ary.
  # Multi-line
  # comment.
  slice(5..-1)&.
  first&.
  zero?
```

# Commenting out lines

```ruby
ary \
  #.slice(5..-1) \
  &.first \
  &.zero?

#ary \
#  .slice(5..-1) \
#  &.first \
#  #&.zero?

#ary \
#  .slice(5..-1) \
#  &.first \
#  #&.zero?

ary.
  #slice(5..-1)&.
  first&.
  zero?

#ary.
#  slice(5..-1)&.
#  first&.
#  #zero?

#ary.
#  #slice(5..-1)&.
#  #first&.
#  #zero?

ary
  #.slice(5..-1)
  &.first
  &.zero?

ary
  slice(5..-1)
  &.first
  #&.zero?

ary
  #slice(5..-1)
  #&.first
  #&.zero?
```

As for "trailing `\\` leading `.`" and "trailing `.`": Notice the inconsistency:
iff you comment out the last line, things will break.

# REPL pasting

```ruby
ary \
  .slice(5..-1) \
  &.first \
  &.zero?
```

```ruby
ary.
  slice(5..-1)&.
  first&.
  zero?
```

```ruby
ary
  .slice(5..-1)
  &.first
  &.zero?
```

In this, `1 < m < n < 4` (and as for `Array#[]`, "`4 = -1`").

## Things you could do and why

The "whys" are just examples.

### Pasting line 1

What's that `ary` thingy?

### Pasting line 1..m:

It fails at some point, let me divide and conquer to see some intermediate
value.

### Pasting line m..-1:

I have this toy `ary` at hand in my REPL, modify this the same way as would
happen with the more complex one that fails.

### Pasting line m..n:

As above, but divide and conquer to understand what is happening.

### Pasting line m..m:

What's this thing doing to my toy `ary`?


## Support matrix

```
                        1       1..m    m..-1   m..n    m..m
                        -----   -----   -----   -----   -------
trailing \, leading .   + (a)   + (a)   +       + (a)   + (a)
           trailing .   + (b)   + (b)   + (c)   + (c)   + (b,c)
            leading .   +       -       -       -       +
```

(a): After the trailing `\\`, hit enter (again).
(b): Delete the trailing `.`.
(c): Add a `.` in the REPL before pasting.

Reply:
Just updated the above.

I think that it's more important to have comments in code (and that it's possible to comment the code "freely") than it is to be able to debug code easily (by either commenting out lines or C&P'ing to a REPL).
Reasons:
* The comments are supposed to exist for a long time and be read multiple times (by a later self)
* Debugging is (hopefully) done less frequently.
* It's not too hard to also comment out a trailing `\` or `.`.
* It's not too hard to add a `.` or delete a trailing `\` or `.` in a REPL.

Thus, I think I still lean towards "trailing `.`". But!

Assume you care more about commenting out code than pasting in a REPL. Then:

> Why not get the benefits from "leading `.`" if you don't need the "comment freedom"?

"Because consistency" and "your eyes should only care about one" are both rather awful reasons IMO. Seasoned developers can for sure easily read both.

My thinking: Stop commenting out code! Drop in `binding.pry` on the line before, and paste into the REPL. That's usually a far superior way to understand code.

Thus, "trailing `.`" is the winner for me.

Reply:
From the above, I would agree that trailing `.` is preferable.

However, visually and semantically, I still much prefer a leading `.`.
In a sense, I read a leading `.` as a self-sufficient line, you start somewhere and you know where you're going.
I read the `.` as something like _and then do_. I read
```
ary
  .slice(5..-1)
  &.first
  &.zero?
```
as 
> Take `ary`,
>   then `slice` it,
>   then `first`,
>   then `zero?`

I read
```
ary. # Single-line comment.
  slice(5..-1)&.
  first&.
  zero?
```
as
> Take `ary` then,
>   `slice` it then,
>   `first` then,
>   `zero?`

Where I clearly prefer the former.

With regards to:
> Leading dots are an antipattern:
>
> 1. They require the continuation to be present at precisely next line. (I.e. you cannot put more than one in-line comment in between.)
> 2. They break mental context by requiring one to always analyze one more line of code.
> 3. Most importantly, code written with leading dots cannot be copied and pasted to irb/pry.

1. That's true. But I don't see that as a major obstacle. If you're chaining complex enough functions
    that you feel like you need to add multiline comments between them, you could probably refactor.
2. They don't break mental context, the indentation on the next line indicate if it continues or not, at the very worst you can look at the first character.
3. True, but when I'm in _ViM_ I can just select the lines and `shift + j` to join them.

I think the most important part is that we're consistent about our choice. So I'd gladly use trailing `.`s :)